kv/kvclient/rangefeed: TestUnrecoverableErrors failed

kv/kvclient/rangefeed.TestUnrecoverableErrors [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9163467?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9163467?buildTab=artifacts#/) on master @ [b89fa2cc4bc1fb9447ab1009b34b6f354f9618f0](https://github.com/cockroachdb/cockroach/commits/b89fa2cc4bc1fb9447ab1009b34b6f354f9618f0):


```
I230321 07:08:43.109842 3636474 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r36/1:/Table/3{4-5}] 1866  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:43.757693 2516882 testutils/soon.go:41  [-] 1867  SucceedsSoon: expected internal error
I230321 07:08:43.841709 2541406 jobs/adopt.go:112  [T1,n1] 1868  claimed 2 jobs
I230321 07:08:43.863828 3662753 jobs/adopt.go:261  [T1,n1] 1869  job 102: resuming execution
I230321 07:08:43.863886 3662751 jobs/adopt.go:261  [T1,n1] 1870  job 100: resuming execution
I230321 07:08:43.864392 3662752 jobs/adopt.go:261  [T1,n1] 1871  job 101: resuming execution
E230321 07:08:43.898730 3662753 jobs/adopt.go:225  [T1,n1] 1872  could not run claimed job 102: no resumer is available for AUTO CONFIG RUNNER
I230321 07:08:43.954611 3667907 jobs/registry.go:1494  [T1,n1] 1873  POLL JOBS STATS job 101: stepping through state running
I230321 07:08:43.963774 3668375 jobs/registry.go:1494  [T1,n1] 1874  KEY VISUALIZER job 100: stepping through state running
I230321 07:08:44.108369 3683472 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r57/1:/Table/5{6-7}] 1875  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:44.758398 2516882 testutils/soon.go:41  [-] 1876  SucceedsSoon: expected internal error
I230321 07:08:45.109292 3716057 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r20/1:/Table/1{8-9}] 1877  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:45.759213 2516882 testutils/soon.go:41  [-] 1878  SucceedsSoon: expected internal error
I230321 07:08:46.760375 2516882 testutils/soon.go:41  [-] 1879  SucceedsSoon: expected internal error
I230321 07:08:47.113128 3773482 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r21/1:/Table/{19-20}] 1880  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:47.357576 2523786 kv/kvserver/replica_rangefeed.go:703  [T1,n1,s1,r34/1:/Table/3{2-3}] 1881  RangeFeed closed timestamp 1679382526.664259796,0 is behind by 693.299357ms
I230321 07:08:47.760615 2516882 testutils/soon.go:41  [-] 1882  SucceedsSoon: expected internal error
I230321 07:08:48.113674 3798625 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r3/1:/System/{NodeLive…-tsd}] 1883  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:48.761617 2516882 testutils/soon.go:41  [-] 1884  SucceedsSoon: expected internal error
I230321 07:08:49.085757 2524347 2@util/log/event_log.go:32  [T1,n1] 1885 ={"Timestamp":1679382529085753139,"EventType":"runtime_stats","MemRSSBytes":343293952,"GoroutineCount":371,"MemStackSysBytes":6455296,"GoAllocBytes":118848824,"GoTotalBytes":233693528,"GoStatsStaleness":0.001443705,"HeapFragmentBytes":75195080,"HeapReservedBytes":163840,"HeapReleasedBytes":185212928,"CGoAllocBytes":9595936,"CGoTotalBytes":28246016,"CGoCallRate":0.19998817,"CPUUserPercent":56.99663,"CPUSysPercent":3.8997695,"GCPausePercent":0.00540516,"GCRunCount":71,"NetHostRecvBytes":104390,"NetHostSendBytes":104390}
I230321 07:08:49.086035 2524347 2@server/status/runtime_log.go:47  [T1,n1] 1886  runtime stats: 327 MiB RSS, 371 goroutines (stacks: 6.2 MiB), 113 MiB/223 MiB Go alloc/total(stale) (heap fragmentation: 72 MiB, heap reserved: 160 KiB, heap released: 177 MiB), 9.2 MiB/27 MiB CGO alloc/total (0.2 CGO/sec), 57.0/3.9 %(u/s)time, 0.0 %gc (71x), 102 KiB/102 KiB (r/w)net
I230321 07:08:49.115881 3826158 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r59/1:/Table/5{8-9}] 1887  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:49.150564 3829294 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r18/1:/Table/1{6-7}] 1888  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:49.761796 2516882 testutils/soon.go:41  [-] 1889  SucceedsSoon: expected internal error
I230321 07:08:50.117210 3855942 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r56/1:/Table/5{5-6}] 1890  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:50.762632 2516882 testutils/soon.go:41  [-] 1891  SucceedsSoon: expected internal error
I230321 07:08:51.762901 2516882 testutils/soon.go:41  [-] 1892  SucceedsSoon: expected internal error
I230321 07:08:52.127093 3911366 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r27/1:/Table/2{5-6}] 1893  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:52.763041 2516882 testutils/soon.go:41  [-] 1894  SucceedsSoon: expected internal error
I230321 07:08:53.619271 2523786 kv/kvserver/replica_rangefeed.go:703  [T1,n1,s1,r34/1:/Table/3{2-3}] 1895  RangeFeed closed timestamp 1679382533.117193795,0 is behind by 502.055495ms
I230321 07:08:53.763698 2516882 testutils/soon.go:41  [-] 1896  SucceedsSoon: expected internal error
I230321 07:08:54.126417 3965677 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r10/1:/Table/{6-7}] 1897  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:54.763856 2516882 testutils/soon.go:41  [-] 1898  SucceedsSoon: expected internal error
I230321 07:08:55.764720 2516882 testutils/soon.go:41  [-] 1899  SucceedsSoon: expected internal error
I230321 07:08:56.131800 4016976 13@kv/kvserver/replicate_queue.go:842  [T1,n1,replicate,s1,r47/1:/Table/4{5-6}] 1900  error processing replica: 0 of 1 live stores are able to take a new replica for the range (1 already has a voter, 0 already have a non-voter); likely not enough nodes in cluster
I230321 07:08:56.765184 2516882 testutils/soon.go:41  [-] 1901  SucceedsSoon: expected internal error
I230321 07:08:57.765855 2516882 testutils/testcluster/testcluster.go:144  [-] 1902  TestCluster quiescing nodes
    rangefeed_external_test.go:906: condition failed to evaluate within 45s: expected internal error
I230321 07:08:57.766794 2521136 server/start_listen.go:103  [T1,n1] 1903  server shutting down: instructing cmux to stop accepting
W230321 07:08:57.767681 2541396 jobs/registry.go:1052  [T1,n1] 1905  canceling all adopted jobs due to stopper quiescing
I230321 07:08:57.767882 3668375 jobs/registry.go:1494  [T1,n1] 1906  KEY VISUALIZER job 100: stepping through state succeeded
I230321 07:08:57.768623 2628359 rpc/context.go:2257  [T1,n1,rnode=1,raddr=127.0.0.1:44403,class=default,rpc] 1907  connection heartbeat loop ended with err: <nil>
W230321 07:08:57.767609 2536853 sql/sqlliveness/slinstance/slinstance.go:312  [n1] 1904  exiting heartbeat loop
W230321 07:08:57.768921 2536853 sql/sqlliveness/slinstance/slinstance.go:297  [n1] 1908  exiting heartbeat loop
I230321 07:08:57.768265 2541139 sql/stats/automatic_stats.go:507  [T1,n1] 1909  quiescing auto stats refresher
E230321 07:08:57.769017 2536853 server/server_sql.go:493  [n1] 1910  failed to run update of instance with new session ID: node unavailable; try another peer
E230321 07:08:57.769863 3667907 jobs/registry.go:924  [T1,n1] 1911  error getting live session: node unavailable; try another peer
I230321 07:08:57.770692 2566852 jobs/registry.go:1494  [T1,n1] 1912  AUTO SPAN CONFIG RECONCILIATION job 849713853864673281: stepping through state succeeded
W230321 07:08:57.780338 3668375 jobs/adopt.go:511  [T1,n1] 1913  could not clear job claim: clear-job-claim: failed to send RPC: sending to all replicas failed; last error: ba: Scan [/Table/15/1/100,/Table/15/1/101), [txn: 1db1448f], [can-forward-ts] RPC error: node unavailable; try another peer
--- FAIL: TestUnrecoverableErrors (48.91s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnrecoverableErrors.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25700

--- a/pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go
@@ -106,14 +106,14 @@ func muxRangeFeed(
 // the entire stream must be torn down, and all active range feeds should be
 // restarted.
 type muxStream struct {
-	nodeID  roachpb.NodeID
-	streams syncutil.IntMap // streamID -> *activeMuxRangeFeed.
+	nodeID roachpb.NodeID
 
 	// mu must be held when starting rangefeed.
 	mu struct {
 		syncutil.Mutex
-		sender rangeFeedRequestSender
-		closed bool
+		sender  rangeFeedRequestSender
+		streams map[int64]*activeMuxRangeFeed
+		closed  bool
 	}
 }
 
@@ -311,6 +311,7 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
 
 	ms := muxStream{nodeID: nodeID}
 	ms.mu.sender = mux
+	ms.mu.streams = make(map[int64]*activeMuxRangeFeed)
 	if err := future.MustSet(stream, muxStreamOrError{stream: &ms}); err != nil {
 		return err
 	}
@@ -330,7 +331,7 @@ func (m *rangefeedMuxer) startNodeMuxRangeFeed(
 			recvErr = nil
 		}
 
-		return ms.closeWithRestart(ctx, recvErr, func(_ int64, a *activeMuxRangeFeed) error {
+		return ms.closeWithRestart(ctx, recvErr, func(a *activeMuxRangeFeed) error {
 			return m.restartActiveRangeFeed(ctx, a, recvErr)
 		})
 	}
@@ -399,7 +400,7 @@ func (m *rangefeedMuxer) receiveEventsFromNode(
 			if active.catchupRes != nil {
 				m.ds.metrics.RangefeedErrorCatchup.Inc(1)
 			}
-			ms.streams.Delete(event.StreamID)
+			ms.deleteStream(event.StreamID)
 			if err := m.restartActiveRangeFeed(ctx, active, t.Error.GoError()); err != nil {
 				return err
 			}
@@ -421,14 +422,10 @@ func (m *rangefeedMuxer) receiveEventsFromNode(
 		// get stuck in the first place.
 		if timeutil.Now().Before(nextStuckCheck) {
 			if threshold := stuckThreshold(); threshold > 0 {
-				if _, err := ms.eachStream(func(id int64, a *activeMuxRangeFeed) error {
-					if !a.startAfter.IsEmpty() && timeutil.Since(a.startAfter.GoTime()) > stuckThreshold() {
-						ms.streams.Delete(id)
-						return m.restartActiveRangeFeed(ctx, a, errRestartStuckRange)
+				for _, a := range ms.purgeStuckStreams(threshold) {
+					if err := m.restartActiveRangeFeed(ctx, a, errRestartStuckRange); err != nil {
+						return err
 					}
-					return nil
-				}); err != nil {
-					return err
 				}
 			}
 			nextStuckCheck = timeutil.Now().Add(stuckCheckFreq())
@@ -491,29 +488,46 @@ func (c *muxStream) startRangeFeed(
 		return err
 	}
 
-	// mu must be held while marking this stream in flight (streams.Store) to
-	// synchronize with mux termination. When node mux terminates, it invokes
-	// c.closeWithRestart(), which marks this mux stream connection closed and
-	// restarts all active streams.  Thus, we must make sure that this streamID
-	// gets properly recorded even if mux go routine terminates right after the
-	// above sender.Send() succeeded.
-	c.streams.Store(streamID, unsafe.Pointer(stream))
+	// As soon as we issue Send above, the stream may return an error that
+	// may be seen by the event consumer (receiveEventsFromNode).
+	// Therefore, we update streams map under the lock to ensure that the
+	// receiver will be able to observe this stream.
+	c.mu.streams[streamID] = stream
 	return nil
 }
 
-func (c *muxStream) lookupStream(streamID int64) *activeMuxRangeFeed {
-	v, ok := c.streams.Load(streamID)
-	if ok {
-		return (*activeMuxRangeFeed)(v)
+func (c *muxStream) lookupStream(streamID int64) (a *activeMuxRangeFeed) {
+	c.mu.Lock()
+	a = c.mu.streams[streamID]
+	c.mu.Unlock()
+	return a
+}
+
+func (c *muxStream) purgeStuckStreams(threshold time.Duration) (stuck []*activeMuxRangeFeed) {
+	c.mu.Lock()
+	for streamID, a := range c.mu.streams {
+		if !a.startAfter.IsEmpty() && timeutil.Since(a.startAfter.GoTime()) > threshold {
+			stuck = append(stuck, a)
+			delete(c.mu.streams, streamID)
+		}
 	}
-	return nil
+	c.mu.Unlock()
+	return stuck
+}
+
+func (c *muxStream) deleteStream(streamID int64) {
+	c.mu.Lock()
+	delete(c.mu.streams, streamID)
+	c.mu.Unlock()
 }
 
 func (c *muxStream) closeWithRestart(
-	ctx context.Context, reason error, restartFn func(streamID int64, a *activeMuxRangeFeed) error,
+	ctx context.Context, reason error, restartFn func(a *activeMuxRangeFeed) error,
 ) error {
 	c.mu.Lock()
 	c.mu.closed = true
+	toRestart := c.mu.streams
+	c.mu.streams = nil
 	c.mu.Unlock()
 
 	// make sure that the underlying error is not fatal. If it is, there is no
@@ -522,22 +536,14 @@ func (c *muxStream) closeWithRestart(
 		return err
 	}
 
-	n, err := c.eachStream(restartFn)
-	if log.V(1) {
-		log.Infof(ctx, "mux to node %d restarted %d streams: err=%v", c.nodeID, n, err)
+	for _, a := range toRestart {
+		if err := restartFn(a); err != nil {
+			return err
+		}
 	}
-	return err
-}
 
-// eachStream invokes provided function for each stream.  If the function
-// returns an error, iteration stops.  Returns number of streams processed.
-func (c *muxStream) eachStream(
-	fn func(streamID int64, a *activeMuxRangeFeed) error,
-) (n int, err error) {
-	c.streams.Range(func(key int64, value unsafe.Pointer) bool {
-		err = fn(key, (*activeMuxRangeFeed)(value))
-		n++
-		return err == nil
-	})
-	return n, err
+	if log.V(1) {
+		log.Infof(ctx, "mux to node %d restarted %d streams", c.nodeID, len(toRestart))
+	}
+	return nil
 }

